RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClustersFromSoAProducer.cc

```diff
@@ -0,0 +1,165 @@
+    for (int32_t i = 0; i < soaCellsOut_v.metadata().size(); ++i) {
+      if (soaCellsOut_v[i].clusterIndex() == -1) {
+        continue;
+      }
+      assert(soaCellsOut_v[i].clusterIndex() < (int)clusters->size());
+      (*clusters)[soaCellsOut_v[i].clusterIndex()].addHitAndFraction(soaCells_v[i].detid(), 1.f);
+      if (soaCells_v[i].timeError() < 0.f) {
+        continue;
+      }
+      times[soaCellsOut_v[i].clusterIndex()].push_back(soaCells_v[i].time());
+      timeErrors[soaCellsOut_v[i].clusterIndex()].push_back(1.f /
+                                                            (soaCells_v[i].timeError() * soaCells_v[i].timeError()));
+    }
+
+    // Finally, compute and assign the time to each cluster.
+    std::vector<std::pair<float, float>> cluster_times;
+    hgcalsimclustertime::ComputeClusterTime timeEstimator;
+    for (unsigned i = 0; i < clusters->size(); ++i) {
+      if (detector_ != "BH") {
+        cluster_times.push_back(timeEstimator.fixSizeHighestDensity(times[i], timeErrors[i], hitsTime_));
+      } else {
+        cluster_times.push_back(std::pair<float, float>(-99.f, -1.f));
+      }
+    }
+
+#if DEBUG_CLUSTERS_ALPAKA
+    //      hgcalUtils::DumpCellsSoA dumperCellsSoA;
+    //      dumperCellsSoA.dumpInfos(cells);
+
+    hgcalUtils::DumpClusters dumper;
+    dumper.dumpInfos(*clusters, true);
+
+//      hgcalUtils::DumpClustersSoA dumperSoA;
+//      dumperSoA.dumpInfos(clustersSoA);
+#endif
+
+    auto clusterHandle = iEvent.put(std::move(clusters));
+
+    auto timeCl = std::make_unique<edm::ValueMap<std::pair<float, float>>>();
+    edm::ValueMap<std::pair<float, float>>::Filler filler(*timeCl);
+    filler.insert(clusterHandle, cluster_times.begin(), cluster_times.end());
+    filler.fill();
+    iEvent.put(std::move(timeCl), timeClname_);
+
+    // The layerClusterMask for the HGCAL detector is created at a later
+    // stage, when the layer clusters from the different components of HGCAL
+    // are merged together into a unique collection. For the case of HFNose,
+    // since there is no further merging step needed, we create the
+    // layerClustersMask directly here.
+    if (detector_ == "HFNose") {
+      std::unique_ptr<std::vector<float>> layerClustersMask(new std::vector<float>);
+      layerClustersMask->resize(clusterHandle->size(), 1.0);
+      iEvent.put(std::move(layerClustersMask), "InitialLayerClustersMask");
+    }
+  }
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("srcDevice", edm::InputTag("hgCalSoALayerClustersProducer"));
+    desc.add<edm::InputTag>("hgcalRecHitsLayerClustersSoA", edm::InputTag("TO BE DEFINED"));
+    desc.add<edm::InputTag>("hgcalRecHitsSoA", edm::InputTag("TO BE DEFINED"));
+    desc.add<unsigned int>("nHitsTime", 3);
+    desc.add<std::string>("timeClname", "timeLayerCluster");
+    desc.add<std::string>("detector", "EE");
```

Comment:
Can you use `edm::allowedValues` to specify the list of valid values for this option ?
See https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Allow_only_specific_values .